File: HOT100/HOT100p1/3.滑动窗口/438.找到字符串中所有字母异位词.cpp
/*
 * @lc app=leetcode.cn id=438 lang=cpp
 *
 * [438] 找到字符串中所有字母异位词
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int i = 0, j = 0;
        array<int, 26> hash_p{};
        vector<int> res;
        for (auto c : p)
        {
            hash_p[c - 'a'] ++;
        }
        for (; j < s.size(); j ++ )
        {
            hash_p[s[j] - 'a'] --;
            while (hash_p[s[j] - 'a'] < 0)
            {
                hash_p[s[i ++] - 'a'] ++;
            }
            if (j - i + 1 == p.size()) res.push_back(i);
        }
        return res;
    }
};
// @lc code=end
class Solution {
    public:
        vector<int> findAnagrams(string s, string p) {
            int i = 0, j = 0;
            array<int, 26> hash_p{}, hash_s{};
            vector<int> res;
            for (auto c : p)
            {
                hash_p[c - 'a'] ++;
            }
            while (j < s.size())
            {
                hash_s[s[j] - 'a'] ++;
                if (j - i + 1 == p.size())
                {
                    if (hash_p == hash_s) res.push_back(i);
                    hash_s[s[i ++] - 'a'] --;
                }
                j ++;
            }
            return res;
        }
    };

File: HOT100/HOT100p5/post/438.找到字符串中所有字母异位词.py
#
# @lc app=leetcode.cn id=438 lang=python3
#
# [438] 找到字符串中所有字母异位词
#
from mytools import *
# @lc code=start
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        h = Counter(p)
        need = len(h)
        j = 0
        res = []
        for i in range(len(s)):
            if s[i] in h:
                h[s[i]] -= 1
                if h[s[i]] == 0: need -= 1
            while need == 0:
                if i - j + 1 == len(p):
                    res.append(j)
                if s[j] in h:
                    h[s[j]] += 1
                    if h[s[j]] > 0: need += 1
                j += 1
        return res
# @lc code=end

